enhancements/cluster-high-availability-mode-api.md

```diff
@@ -0,0 +1,360 @@
+field (or a few fields) to the existing API, because it requires more
+code to be updated, including having the RBAC rules configured to
+allow the service accounts to see it.
+
+### Expose the cluster profile name directly
+
+We could add a `clusterProfile` field to an existing or new
+configuration API. This would require each operator to interpret each
+new profile, which would make maintaining profiles even more work than
+it is now. It may also encourage us to create more profiles, when we
+have stated elsewhere that our goal is to have as few as possible.
+
+### Use the cluster profile to have different Deployments for operators
+
+We could use different manifests to have the cluster-version-operator
+deploy operators with different configuration (via command line
+arguments or environment variables). This would give us a way to
+configure operators differently for each profile, but those settings
+would be less discoverable and we would not be able to offer users a
+way to change them at runtime (if we ever want to do that).
+
+### Recognize HA/non-HA configuration based on control plane node count
+
+Operators could look at the number of control plane nodes and use that
+information to determine whether HA is enabled (3 or more nodes) or
+not (1 node).
+
+This would require giving the operator read access to Node resources,
+and would be subject to race conditions as the control plane
+forms. For example, the cluster-etcd-operator [requires 3 control
+plane
+nodes](https://github.com/openshift/cluster-etcd-operator/blob/98590e6ecfe282735c4eff01432ae40b29f81202/pkg/etcdenvvar/etcd_env.go#L72)
+to exist before deploying its operand. If it counted nodes, it could
+not tell the difference between a cluster starting to form with 1
+node, then more, or a cluster that is only ever going to include 1
+node.
+
+Looking at nodes also will not work in clusters using an external
+control plane, since there will not be any special control plane
+nodes.
+
+### Use operator-specific APIs to manage the behaviors
+
+The cluster-etcd-operator and cluster-authentication-operator both
+include flags to enter an unsupported non-HA operating mode. We could
+change those APIs to make the flags supported, and add similar flags
+to other components, such as cluster-ingress-operator and
+machine-config-operator. This would offer more flexibility, since each
+component could be configured separately. However, these settings are
+needed during cluster bootstrapping, so this approach would tie the
+installer to the APIs for each of the operators so they could be
+configured separately.
+
+Complexity will grow further if we add capabilities for managing the
+resource consumption of operators and their operands, since we risk
+having different ways to express those constraints in different APIs.
+
+### Manage upgrades by tracking settings in each consuming operator
+
+When we make the API mutable we will need to protect against older
```

Comment:
I'm not sure what you mean by "misclassification"?

Reply:
If we scale from 3 nodes to 1 or vice versa the cluster will try to enforce an invalid HA mode (misclassified) for the cluster topology. Sometime soon we will need to support 1 node and 3 node clusters. Will it not be supported for a single node UPI cluster to scale up to 3? Because the HA mode is static operators will try to enforce HA mode that is not enforceable given the node count.

failure case: single node cluster scales to 3, then upgrades. But because it was created as a single node it does not have a quorum-guard. Cluster loses quorum during upgrade. Who owns that bug?

It seems like this infra controller should go degraded saying HA mode not met. But in this case customer doesn't want `None` they want `Full`. So it seems that until such time that HA Mode is Mutable or a controller exists that will update status based on topology scaling from 1 to 3 nodes is not supported?

Reply:
OK, I think I see what you're saying.

If a single-node deployment scales the _control plane_ up then I would expect the relevant operators to scale their operands, including deploying quorum guard for etcd. That implies this API needs to be tied to the one that scales the control plane, somehow. I don't think we have concrete plans to handle this case, but we can make a note that it's going to take significant thought to work out implications like this.

We do have plans to deal with the case where a single-node deployment gains some worker nodes. In that configuration, I wouldn't expect a change in the control plane operands because those worker nodes shouldn't run the control plane components.

Reply:
> If a single-node deployment scales the control plane up then I would expect the relevant operators to scale their operands, including deploying quorum guard for etcd. 

thanks for the comments. this is my point I guess. The value of the API is arbitrary post-install so the operator can not trust its value to reconcile against as we must still count nodes to verify our actions. These actions "quorum-guard deployment replica 3"  would go against the expectations of HA Mode  `None`.

I understand more work is required but if we can consider this case moving forward I am happy. thanks

Reply:
I've added this to the risks section above around line 158. Please let me know if I captured your concerns accurately there.